### ISSUE TITLE
Consider using poetry

### ISSUE THREAD

**Issue by Yajo** [odoo/odoo#63138]
Impacted versions: all, although I guess you would only fix it on master

Steps to reproduce:

1. Go to https://github.com/odoo/odoo/blob/49b8f1bfde1792079c0fd9cdb218eab20f344be3/requirements.txt
1. See how old & strict these requirements are.
1. Read https://pip.pypa.io/en/stable/user_guide/#changes-to-the-pip-dependency-resolver-in-20-3-2020
1. Imagine yourself having to install an external library without breaking those old & strict dependencies.

Current behavior:  😿 

Expected behavior: Odoo should update dependencies a bit more often. The best would be after doing a new release, so there's a whole new dev year to test those updates. Maintaining a `requirements.txt` file by hand is quite hard and cumbersome. There are much more modern and easier tools to do that. One of them is https://python-poetry.org/, which under my experience quite well honors its name. This tool allows you to specify version constraints in a less strict and more comfortable way, while still maintaining a lockfile that allows your CI/CD to use a very strict dependency set. Another such tool is https://pipenv.pypa.io/en/stable/ although I've had some bad times with it. Stuff like https://github.com/odoo/odoo/issues/59980 would be way easier to maintain.

cc @xmo-odoo

**Comment by pedrobaeza** [comment]
I suppose @d-fence will have any comment about this

**Comment by Yenthe666** [comment]
One thing I can agree on: dependencies are indeed often quite old and behind.

**Comment by d-fence** [comment]
There is a reason why the dependencies may look old: we want to stick with the current Debian (and Odoo) stable version.
If our requirements are higher that what Debian provides, it will endanger two things:

-  [Debian packaging ](https://packages.debian.org/sid/odoo)
- The users who wants to install Odoo on a stable Debian (or Ubuntu) system without having a mess in their Python libs or who don't want to use python virtual env's

I admit that it's hard to maintain and have it working on different systems but we are working on that.
I'm not sure that Poetry will take the Debian (Ubuntu) python packages into account.

**Comment by d-fence** [comment]
Also, have a look at [this conversation](https://github.com/odoo/runbot/issues/414) to have a better overview of the problem.

**Comment by Yajo** [comment]
The debian version decision had way more sense back in the day when that was the common way to deploy, and also where everybody was using python 2.7 because there were not going to be any other python 2.x release.

I think it's no longer the case. These days, with odoo.sh, poetry, venv, python 3 (with several supported stable versions), docker, guix, k8s... Possibly most of us will not care so much about what's debian's stable version of stuff other than maybe python itself.

I suggest to reconsider that policy, for v15 or later. I think that the best you can do to actually support a platform is test it. A good theoretical roadmap would be:

1. Have a tool that allows you to work sanely with dependencies, such as poetry.
1. Actually have CI on the several python versions enabled. These days, doing that kind of matrixes with github actions, even across windows/mac/linux, is free and easy.

All sounds easier than it is, but this would be very beneficial IMHO.

>  I'm not sure that Poetry will take the Debian (Ubuntu) python packages into account.

There's no relation between poetry and debian whatsoever. You can, however, use poetry to manage versions more comfortably, under the criteria you choose, which can be to follow debian released packages.

**Comment by ronozoro** [comment]
I hope this would move forward, it would be very nice to have.

**Comment by ronozoro** [comment]
> I hope this would move forward, it would be very nice to have.

**Comment by ronozoro** [comment]
> The debian version decision had way more sense back in the day when that was the common way to deploy, and also where everybody was using python 2.7 because there were not going to be any other python 2.x release.
> 
> I think it's no longer the case. These days, with odoo.sh, poetry, venv, python 3 (with several supported stable versions), docker, guix, k8s... Possibly most of us will not care so much about what's debian's stable version of stuff other than maybe python itself.
> 
> I suggest to reconsider that policy, for v15 or later. I think that the best you can do to actually support a platform is test it. A good theoretical roadmap would be:
> 
> 1. Have a tool that allows you to work sanely with dependencies, such as poetry.
> 2. Actually have CI on the several python versions enabled. These days, doing that kind of matrixes with github actions, even across windows/mac/linux, is free and easy.
> 
> All sounds easier than it is, but this would be very beneficial IMHO.
> 
> > I'm not sure that Poetry will take the Debian (Ubuntu) python packages into account.
> 
> There's no relation between poetry and debian whatsoever. You can, however, use poetry to manage versions more comfortably, under the criteria you choose, which can be to follow debian released packages.


I hope this would move forward, it would be very nice to have, and well done explanation

**Comment by AltaherPresto** [comment]
Any progress or updates on moving package requirements to `Poetry`?

**Comment by rafalkrupinski** [comment]
And poetry isn't even the only choice any more.
Just moving to `pyproject.toml` would be a progress

**Comment by yajo** [comment]
There's a proposal in https://github.com/odoo/odoo/pull/44001. Open for 4 years now, though...

**Comment by yajo** [comment]
> `uv` is the new "best thing"
> [...]
> simply moving to `pyproject.toml` would be an achviement

FWIW poetry is both a build backend and a package manager. `uv`, however, is not a build backend; just a package manager. This means that one can use any build backend with `uv`, sa long as it has a valid `pyproject.toml` file.

In #44001 there's a working minimal `pyproject.toml`. It's not related to poetry at all, but it allows working with `uv` or many other systems correctly. I'm using it for production for a lot of time, and works perfectly.

Also, for the odoo/enterprise and odoo/design-themes repos, you could just use `hatch-odoo` for packaging. It also works perfectly there.

In short, the needed effort is minimal, but the benefits are quite big for downstream packaging. I hope this gets fixed at some point. I changed the title fwiw.
